ksp_plugin_test/interface_compatibility_test.cpp

```diff
@@ -0,0 +1,124 @@
+﻿
+#include <string>
+#include <vector>
+
+#include "base/file.hpp"
+#include "base/pull_serializer.hpp"
+#include "base/push_deserializer.hpp"
+#include "glog/logging.h"
+#include "gmock/gmock.h"
+#include "gtest/gtest.h"
+#include "ksp_plugin/interface.hpp"
+#include "ksp_plugin/plugin.hpp"
+#include "serialization/ksp_plugin.pb.h"
+#include "testing_utilities/serialization.hpp"
+
+namespace principia {
+namespace interface {
+
+using base::OFStream;
+using base::PullSerializer;
+using base::PushDeserializer;
+using ksp_plugin::Plugin;
+using testing_utilities::ReadLinesFromBase64File;
+using testing_utilities::ReadLinesFromHexadecimalFile;
+using ::testing::NotNull;
+
+const char preferred_compressor[] = "gipfeli";
+const char preferred_encoder[] = "base64";
+
+class InterfaceCompatibilityTest : public testing::Test {
+ protected:
+  void CheckSaveCompatibility(std::filesystem::path const& filename,
+                              std::string_view const compressor,
+                              std::string_view const encoder) {
+    Plugin const* plugin = nullptr;
+
+    // Read a plugin from a file containing only the "serialized_plugin = "
+    // lines, with "serialized_plugin = " dropped.
+    {
+      PushDeserializer* deserializer = nullptr;
+      auto const lines =
+          encoder == "hexadecimal" ? ReadLinesFromHexadecimalFile(filename)
+          : encoder == "base64"    ? ReadLinesFromBase64File(filename)
+                                   : std::vector<std::string>{};
+      CHECK(!lines.empty());
+      LOG(ERROR) << "Deserialization starting";
+      for (std::string const& line : lines) {
+        principia__DeserializePlugin(line.c_str(),
+                                     &deserializer,
+                                     &plugin,
+                                     compressor.data(),
+                                     encoder.data());
+      }
+      principia__DeserializePlugin("",
+                                   &deserializer,
+                                   &plugin,
+                                   compressor.data(),
+                                   encoder.data());
+      LOG(ERROR) << "Deserialization complete";
+    }
+    EXPECT_THAT(plugin, NotNull());
+
+    // Write that plugin back to another file with the preferred format.
```

Comment:
Split the reading and writing into separate files.

Reply:
Done.